Refactor `DefaultImapDecoderFactory` to Use POJO Style and Constructo…

superbAura opened the pull request:
…r Injection

- Updated `DefaultImapDecoderFactory` to adhere to POJO design principles.
- Replaced internal `ImapParserFactory` creation with constructor injection of `UnpooledStatusResponseFactory`.
- Improved decoupling and maintainability by explicitly managing dependencies via the constructor.
- Updated class and method documentation to reflect changes and clarify the role of the factory.

This change enhances the clarity and testability of the `DefaultImapDecoderFactory` by removing direct dependency creation and enforcing a cleaner, more modular structure.

superbAura commented:
Hello,

I am currently reviewing Apache James for developing a mail engine. While forking the source, I noticed a simple TO-DO comment and made a commit request.

Starting from 2025, we plan to use Apache James extensively. If given the opportunity, I would like to contribute significantly to its improvements.

Thank you!

Arsnael commented on protocols/imap/src/main/java/org/apache/james/imap/main/DefaultImapDecoderFactory.java:
```diff
@@ -27,20 +27,35 @@
+     *
+     * @param unpooledStatusResponseFactory The factory for creating status responses.
+     */
+    public DefaultImapDecoderFactory(UnpooledStatusResponseFactory unpooledStatusResponseFactory) {
```
That code is very old... Aint very sure about this part but problems I can see:

- several test classes depend on DefaultImapDecoderFactory.buildImapDecoder() . This change without being adapted is definitely gonna break them 
- have a look in `IMAPServerModule` . The DefaultImapDecoder is being provided by building directly it, not passing by the Factory. It would make sense maybe to do a refactoring there as well.
- you might want to pass the ImapCommandParserFactory as an argument too (to use in IMAPServerModule)
- you could maybe have a static method that would return a default imap encoder (as of now) just for tests purposes

superbAura replied:
Understood. I will check if it's possible to refactor based on what you mentioned.
Thank you for the review.

Arsnael replied:
No problem, don't hesitate to ask if you got issues

superbAura replied:
First, added a default constructor for DefaultImapDecoderFactory.

This addition enhances flexibility by allowing other classes to use the default constructor.

Additionally, added another constructor that accepts an implementation of StatusResponseFactory.

This change is to provide flexibility for the ResponseFactory in DefaultImapDecoderFactory in the future.

Also modified two other test classes to use the constructor-based approach

superbAura replied:
The continuous-integration/jenkins/pr-merge failed. I will check it.

Arsnael reviewed (commented):
Hello. Thank you for the contribution, happy to see you so interested in using and contributing more to james in a near future. Welcome!

Comment below

chibenwa reviewed (commented):
THis is great! Happy to further hear about your use cases and welcome onboard the community.

So as this very changeset I cannot say better than @Arsnael ;-)

chibenwa commented on protocols/imap/src/test/java/org/apache/james/imap/main/DefaultImapDecoderFactoryTest.java:
```diff
@@ -0,0 +1,39 @@
+package org.apache.james.imap.main;
```
THis file misses the compulsory ASF V2 license

superbAura replied:
Added the ASF license notice to the source files. Please feel free to review and let me know if there are any concerns. Thank you! !

Arsnael commented on protocols/imap/src/main/java/org/apache/james/imap/main/DefaultImapDecoderFactory.java:
```diff
@@ -27,20 +28,46 @@
+     */
+    public DefaultImapDecoderFactory(StatusResponseFactory statusResponseFactory) {
+        this.statusResponseFactory = statusResponseFactory;
+        this.imapCommandParserFactory = new ImapParserFactory(statusResponseFactory);
```
Not necessary, you should pass that as an argument as well, and not building the default (that you can pass via your `DefaultImapDecoderFactory()` constructor)

Arsnael commented:
3 Checkstyle issues in the build. What IDE are you using btw? i personally use Intellij and it allows to navigate easily in the project. Also you can have a plugin to apply and check the checkstyle (to make sure it respects a certain number of coding rules within the project).

Checkstyle file used by James is at the root f the project: https://github.com/apache/james-project/blob/master/checkstyle.xml

Also I don't see a refactoring in IMAPServerModule to use the factory for building the DefaultImapDecoder?

superbAura commented:
I also use IntelliJ.
@Arsnael 
I will improve the factory and builder pattern in the IMAPServerModule module.

Thank you.

If you have any references that I can refer to, I would appreciate it

Arsnael commented on protocols/imap/src/test/java/org/apache/james/imap/main/DefaultImapDecoderFactoryTest.java:
```diff
@@ -0,0 +1,59 @@
+        // Build an ImapDecoder using the factory and assert it's not null
+        ImapDecoder decoder = factory.buildImapDecoder();
+        assertNotNull(decoder, "ImapDecoder instance should not be null");
+    }
```
Maybe add a unit test for the constructor `public DefaultImapDecoderFactory(ImapCommandParserFactory imapCommandParserFactory, StatusResponseFactory statusResponseFactory)` and I think it should be alright for me

Arsnael replied:
This comment has not been answered yet @superbAura :)

Arsnael commented:
06:36:13,438 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.0:check (check-style) on project protocols-imap: You have 1 Checkstyle violation. -> [Help 1]

Still one checkstyle issue somewhere :)

Arsnael commented:
Why all the indents? Can you not change that please? Thanks

superbAura commented:
@Arsnael 

After applying the Code Style in my IntelliJ, the code style was applied to the source I modified, and the indentation was updated.

I will rollback if it causes any issues.
I fixed the build errors caused by the Code Style. may be .. build runtime is so long

The process of checking Code Style during the build was a refreshing experience for me.
I learned a lot today as well.

superbAura commented:
> Why all the indents? Can you not change that please? Thanks

Shall I rollback and apply all the line breaks?

chibenwa commented:
>  Shall I rollback and apply all the line breaks?

I feel very sorry for this as this is likely to be a big work, but yes please, this change would be way easier to review without all those unrelated changes.

Project policy is to have ll spacings being 4-space based. IDE shall be reconfigured in order to match this.

superbAura commented:
That's a very minor thing. I will edit it and upload it soon

superbAura commented:
> > Shall I rollback and apply all the line breaks?
> 
> I feel very sorry for this as this is likely to be a big work, but yes please, this change would be way easier to review without all those unrelated changes.
> 
> Project policy is to have ll spacings being 4-space based. IDE shall be reconfigured in order to match this.

I have completed the requested changes !

chibenwa reviewed (approved):
Fine with me though a rebase is needed in order to fix the conflict.

superbAura commented:
> Fine with me though a rebase is needed in order to fix the conflict.

Thank you for confirming.

Could you possibly share a backlog or JIRA link for Apache James?

I have a lot of questions regarding Apache James, and if you’re okay with it, could you let me know if there's a Slack channel or another way to ask questions?"

Let me know if you need any changes!

Arsnael commented:
> Could you possibly share a backlog or JIRA link for Apache James?

JIRA backlog: https://issues.apache.org/jira/projects/JAMES/issues/JAMES-3580?filter=allopenissues

> I have a lot of questions regarding Apache James, and if you’re okay with it, could you let me know if there's a Slack channel or another way to ask questions?"

You have two ways:
- subscribing to mailing lists => https://james.apache.org/mail.html
- gitter channel => https://app.gitter.im/#/room/#apache_james-project:gitter.im

superbAura commented:
> > Could you possibly share a backlog or JIRA link for Apache James?
> 
> JIRA backlog: https://issues.apache.org/jira/projects/JAMES/issues/JAMES-3580?filter=allopenissues
> 
> > I have a lot of questions regarding Apache James, and if you’re okay with it, could you let me know if there's a Slack channel or another way to ask questions?"
> 
> You have two ways:
> 
> * subscribing to mailing lists => https://james.apache.org/mail.html
> * gitter channel => https://app.gitter.im/#/room/#apache_james-project:gitter.im

Thankx!

chibenwa commented:
Please

```
#Fetch apache repository
git fetch apache 

git rebase -i apache/master
# squash all the commits into one

git push origin -f HEAD
```

The git conflict is still present:

![image](https://github.com/user-attachments/assets/e77623ff-0cde-41fb-9320-5f1b354b1f58)

superbAura commented:
> Please
> 
> 
> 
> ```
> 
> #Fetch apache repository
> 
> git fetch apache 
> 
> 
> 
> git rebase -i apache/master
> 
> # squash all the commits into one
> 
> 
> 
> git push origin -f HEAD
> 
> ```
> 
> 
> 
> The git conflict is still present:
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/e77623ff-0cde-41fb-9320-5f1b354b1f58)
> 
> 

ok I'll do it soon !!

superbAura commented:
it's done !

@chibenwa

chibenwa reviewed (commented):
This PR reverts some changes unrelated to this work.

Is this what is wanted?

superbAura commented:
> #Fetch apache repository
> 
> git fetch apache 
> 
> 
> 
> git rebase -i apache/master
> 
> # squash all the commits into one
> 
> 
> 
> git push origin -f HEAD



> This PR reverts some changes unrelated to this work.
> 
> Is this what is wanted?

Right. It seems like unrelated changes have been included in this PR, which wasn't intended

It seems this occurred during the process of syncing with the remote and pushing.

In the IMAPServerTest class, it looks like GenericKeyStoreException.class is not being imported correctly. This is causing the current build failure.

Arsnael commented:
Still conflicts... I see some commits from other people that have been merged being pushed here (which is not wanted). 

Cant you just do a rebase on latest master and squash everything into one commit instead of messy merges with master? Thanks :)

Arsnael commented on server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:
```diff
@@ -3305,15 +3305,15 @@ void tearDown() {
+                    })
+                    .threadCount(10)
+                    .operationCount(200)
+                    .runSuccessfullyWithin(Duration.ofMinutes(10));
```
Can you revert all those unnecessary indents in that file please?

superbAura commented:
> Still conflicts... I see some commits from other people that have been merged being pushed here (which is not wanted).
> 
> Cant you just do a rebase on latest master and squash everything into one commit instead of messy merges with master? Thanks :)

I'm not sure if it's going well, but I completed the squash.

Arsnael commented:
Alright the code seems ok now, I will merge and squash all of this on github's end.

Just maybe for next time, on your PR instead of doing numerous git merge with master that lead to absolute commit mess (40+ commits... it makes it hard to review), prefer maybe the git rebase instead?

Like on your local branch :

```
git fetch apache master
git rebase apache/master

[fix conflicts if necessary]

git push origin HEAD -f
```

Would help keep a clean commit history, in sync with master branch, with making life of people reviewing your PR easier.

Thank you for the contribution though, kindly appreciated :)

superbAura commented:
> Alright the code seems ok now, I will merge and squash all of this on github's end.
> 
> Just maybe for next time, on your PR instead of doing numerous git merge with master that lead to absolute commit mess (40+ commits... it makes it hard to review), prefer maybe the git rebase instead?
> 
> Like on your local branch :
> 
> ```
> git fetch apache master
> git rebase apache/master
> 
> [fix conflicts if necessary]
> 
> git push origin HEAD -f
> ```
> 
> Would help keep a clean commit history, in sync with master branch, with making life of people reviewing your PR easier.
> 
> Thank you for the contribution though, kindly appreciated :)

It was a fun experience contributing to open source for the first time. I'll look for another task to work on !
